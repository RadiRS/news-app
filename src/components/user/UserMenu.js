import React, { Component } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import { Animated } from 'react-native';
import Icon from 'react-native-vector-icons/Ionicons';

import { toggleMenu } from '../../stores/actions';

import { Metrics } from '../../themes';
import MenuItem from './UserMenuItem';

export class UserMenu extends Component {
  state = {
    top: new Animated.Value(Metrics.screenHeight)
  };

  componentDidMount() {
    this.handleToggleMenu();
  }

  componentDidUpdate() {
    this.handleToggleMenu();
  }

  handleToggleMenu = () => {
    if (this.props.menu.openMenu) {
      Animated.spring(this.state.top, {
        toValue: 54
      }).start();
    } else {
      Animated.spring(this.state.top, {
        toValue: Metrics.screenHeight
      }).start();
    }
  };

  render() {
    const { user } = this.props;

    return (
      <AnimatedContainer style={{ top: this.state.top }}>
        <Cover>
          <Title>{user.name}</Title>
          <SubTitle>{user.email}</SubTitle>
        </Cover>

        <CloseView onPress={this.props.toggleMenu}>
          <Icon name="ios-close" size={44} color="#546bfb" />
        </CloseView>

        <Content>
          {menus.map((menu, index) => (
            <MenuItem
              key={index}
              icon={menu.icon}
              title={menu.title}
              text={menu.text}
            />
          ))}
        </Content>
      </AnimatedContainer>
    );
  }
}

const Container = styled.View`
  border-radius: 10;
  overflow: hidden;
  position: absolute;
  background: white;
  width: 100%;
  height: 100%;
  z-index: 100;
`;

const AnimatedContainer = Animated.createAnimatedComponent(Container);

const Cover = styled.View`
  height: 142px;
  background: #546bfb;
  justify-content: center;
  align-items: center;
`;

const Title = styled.Text`
  color: white;
  font-size: 24px;
  font-weight: 600;
`;

const SubTitle = styled.Text`
  color: rgba(255, 255, 255, 0.5);
  font-size: 13px;
  margin-top: 8px;
`;

const CloseView = styled.TouchableOpacity`
  height: 44px;
  width: 44px;
  border-radius: 22px;
  background: white;
  justify-content: center;
  align-items: center;
  position: absolute;
  top: 120;
  left: 50%;
  margin-left: -22px;
  z-index: 1;
  elevation: 15;
`;

const Content = styled.View`
  height: ${Metrics.screenHeight};
  background: #f0f3f5;
  padding: 50px;
`;

const mapStateToProps = ({ menu }) => ({
  menu
});

const mapDispatchToProps = {
  toggleMenu
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(UserMenu);

const menus = [
  {
    icon: 'ios-settings',
    title: 'Account',
    text: 'settings'
  },
  {
    icon: 'ios-exit',
    title: 'Log out',
    text: 'See you soon!'
  }
];
