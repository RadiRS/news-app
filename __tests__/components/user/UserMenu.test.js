import 'react-native';
import 'jest-styled-components';
import React from 'react';
import renderer from 'react-test-renderer';
import { UserMenu } from '../../../src/components/user';

test('renders UserMenu component correctly', () => {
  const tree = renderer.create(<UserMenu user={user} menu={menu} />).toJSON();
  expect(tree).toMatchSnapshot();
});

const user = {
  name: '',
  email: ''
};

const menu = {
  openMenu: true
};
