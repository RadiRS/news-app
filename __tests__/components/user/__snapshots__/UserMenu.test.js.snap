// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders UserMenu component correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "borderRadius": 10,
        "height": "100%",
        "overflow": "hidden",
        "position": "absolute",
        "width": "100%",
        "zIndex": 100,
      },
      Object {
        "top": 1333.8537492906403,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#546bfb",
          "height": 142,
          "justifyContent": "center",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "white",
            "fontSize": 24,
            "fontWeight": "600",
          },
        ]
      }
    >
      
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(255, 255, 255, 0.5)",
            "fontSize": 13,
            "marginTop": 8,
          },
        ]
      }
    >
      
    </Text>
  </View>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 22,
        "elevation": 15,
        "height": 44,
        "justifyContent": "center",
        "left": "50%",
        "marginLeft": -22,
        "opacity": 1,
        "position": "absolute",
        "top": 120,
        "width": 44,
        "zIndex": 1,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#546bfb",
            "fontSize": 44,
          },
          undefined,
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#f0f3f5",
          "height": 1334,
          "paddingBottom": 50,
          "paddingLeft": 50,
          "paddingRight": 50,
          "paddingTop": 50,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "marginBottom": 15,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 15,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 32,
              "justifyContent": "center",
              "width": 32,
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#546bfb",
                "fontSize": 24,
              },
              undefined,
              Object {
                "fontFamily": "Ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "paddingLeft": 20,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#3c4560",
                "fontSize": 24,
                "fontWeight": "600",
              },
            ]
          }
        >
          Log out
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#3c4560",
                "fontWeight": "600",
                "marginTop": 5,
                "opacity": 0.6,
              },
            ]
          }
        >
          See you soon!
        </Text>
      </View>
    </View>
  </View>
</View>
`;
